Provide GPS data to AirControlDisplay

Scumi opened the pull request:
Brief summary of the changes
----------------------------

~~- added a short format for the DDMM.MMM location format~~
- added NMEAFormatter to format standard NMEA sentences
- updated the AirControlDisplay driver to forward GPS data via NMEA sentences to the device
- updated the CAILNAV driver to also use the new NMEAFormatter

Related issues and discussions
------------------------------

#170 (previous attempt)
#1112 (recent request)

lordfolken commented on src/Formatter/NMEAFormatter.hpp:
```diff
@@ -0,0 +1,100 @@
+ * <11> Magnetic variation direction, E or W
+ *      (westerly variation adds to true course)
+ */
+bool FormatGPRMC(TCHAR *buffer, size_t size, const NMEAInfo &info);
```
Code-style issue.

Scumi replied:
Ok, will check and update (and also fix the build errors for the tests).

lordfolken commented on src/Form/DigitEntry.cpp:
```diff
@@ -747,6 +753,8 @@ DigitEntry::GetGeoAngle(CoordinateFormat format) const
     break;
 
+  case CoordinateFormat::DDMM_MMM_SHORT:
+
```
code formatting

lordfolken commented on src/Formatter/NMEAFormatter.cpp:
```diff
@@ -0,0 +1,131 @@
+Copyright_License {
+
+  XCSoar Glide Computer - http://www.xcsoar.org/
+  Copyright (C) 2000-2021 The XCSoar Project
```
copyright to current year

lordfolken commented on test/src/TestNMEAFormatter.cpp:
```diff
@@ -0,0 +1,61 @@
+/* Copyright_License {
+
+  XCSoar Glide Computer - http://www.xcsoar.org/
+  Copyright (C) 2000-2022 The XCSoar Project
```
copyright year

lordfolken commented on build/test.mk:
```diff
@@ -486,6 +487,22 @@ TEST_IGC_FILENAME_FORMATTER_SOURCES = \
 
+TEST_NMEA_FORMATTER_SOURCES = \
+	$(SRC)/Units/Descriptor.cpp \
+    $(SRC)/Units/System.cpp \
+    $(SRC)/Atmosphere/AirDensity.cpp \
+    $(SRC)/Device/Parser.cpp \
+    $(SRC)/Device/Driver/FLARM/StaticParser.cpp \
+    $(SRC)/FLARM/Traffic.cpp \
+    $(SRC)/FLARM/FlarmId.cpp \
```
Indents consistentcy

lordfolken commented on src/Device/Driver/AirControlDisplay.cpp:
```diff
@@ -157,6 +163,25 @@ ACDDevice::ParseNMEA(const char *_line, NMEAInfo &info)
+ACDDevice::OnSensorUpdate(const MoreData &basic)
+{
+  NullOperationEnvironment env;
+  char buffer[100];
```
codacy thinks that this can go up one bracket

lordfolken commented on src/Formatter/NMEAFormatter.hpp:
```diff
@@ -0,0 +1,100 @@
+Copyright_License {
+
+  XCSoar Glide Computer - http://www.xcsoar.org/
+  Copyright (C) 2000-2022 The XCSoar Project
```
copyright year

Scumi commented:
Fixed the code style issues and improved on the logic.

I'm not happy with the amount of additions to test.mk, but I'm not sure what a good way around could be. I could create a new library (e.g. "FORMATTERS/libformatters.mk") and just add that as a dependency, which would reduce the clutter a bit.

Scumi commented:
Simplified the logic a bit further and fixed the windows build and general tests.

However, I cannot figure out, why the windows tests fail. If someone could help me out there or give me a hint, that would be well appreciated.

MaxKellermann commented:
Not ready for review as long as unit tests fail

Scumi commented:
> Not ready for review as long as unit tests fail

Build and tests are good now, you can review any time.

Scumi commented:
rebased

MaxKellermann commented on src/Formatter/NMEAFormatter.cpp:
```diff
@@ -0,0 +1,117 @@
+{
+  const GeoPoint location = info.location_available
+    ? info.location
+    : GeoPoint::Zero();
```
`GeoPoint::Zero()` is a valid location that exists on earth. I don't like using "magic" values like that. Can't we just keep the columns empty if we have no location, instead of printing an arbitrary bogus location?

Scumi replied:
While I agree on your remark of "magic" values, I don't know how devices react when the latitude and longitude fields are left empty, together with the N/S/W/E identifiers. Most specifications of the NMEA0183 GPRMC sentence that I found don't mention that case, while a few just set it to 0000.0000. This is of course a valid location, but devices should discard the sentence based on field #2, which sets the position status to valid or invalid.

MaxKellermann commented on src/Formatter/NMEAFormatter.cpp:
```diff
@@ -0,0 +1,117 @@
+/*
+Copyright_License {
+
+  XCSoar Glide Computer - http://www.xcsoar.org/
+  Copyright (C) 2000-2023 The XCSoar Project
+  A detailed list of copyright holders can be found in the file "AUTHORS".
+
+  This program is free software; you can redistribute it and/or
+  modify it under the terms of the GNU General Public License
+  as published by the Free Software Foundation; either version 2
+  of the License, or (at your option) any later version.
+
+  This program is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program; if not, write to the Free Software
+  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
+}
+*/
```
Please update to SPDX headers

Scumi replied:
done

MaxKellermann commented on src/Formatter/NMEAFormatter.cpp:
```diff
@@ -0,0 +1,117 @@
+*/
+
+#include "Units/Units.hpp"
+#include "NMEAFormatter.hpp"
```
A file should include its own header first

Scumi replied:
done

MaxKellermann commented on src/Formatter/NMEAFormatter.cpp:
```diff
@@ -0,0 +1,117 @@
+  const BrokenDateTime now = info.time_available &&
+    info.date_time_utc.IsDatePlausible()
+    ? info.date_time_utc
+    : BrokenDateTime::NowUTC();
```
The fallback to the current time stamp seems surprising and wrong to me, it's not documented, and the system clock may be wrong. Think of a Kobo or a Raspi that has no battery-buffered real-time clock.
I say this library should not mandate any such fallback, it should only take values from its parameters and never do any system calls.

MaxKellermann commented on src/Formatter/NMEAFormatter.cpp:
```diff
@@ -0,0 +1,117 @@
+    ? info.date_time_utc
+    : BrokenDateTime::NowUTC();
+
+  StringFormat(buffer, size,
+               "GPRMC,%02u%02u%02u,%c,%02u%02u.%03u,%c,%03u%02u.%03u,%c,%05.1f,%05.1f,%02u%02u%02u,%05.1f,%c",
+               now.hour, now.minute, now.second,
+               info.location_available ? 'A' : 'V',
+               (int)location.latitude.AbsoluteDegrees(),
+               location.latitude.ToDMM().minutes,
+               location.latitude.ToDMM().decimal_minutes,
+               location.latitude.IsNegative() ? 'S' : 'N',
+               (int)location.longitude.AbsoluteDegrees(),
+               location.longitude.ToDMM().minutes,
+               location.longitude.ToDMM().decimal_minutes,
+               location.longitude.IsNegative() ? 'W' : 'E',
+               (double)Units::ToUserUnit(info.ground_speed, Unit::KNOTS),
+               (double)info.track.Degrees(),
+               now.day, now.month, now.year % 100,
+               info.variation.AbsoluteDegrees(),
+               info.variation.IsNegative() ? 'W' : 'E');
```
Code like this already exists:
https://github.com/XCSoar/XCSoar/blob/3fe70253b7d880b289427307625058aad12da07d/test/src/IGC2NMEA.cpp#L28-L36
Yours is more generic, which is a good thing, but we shouldn't have two implementations.

Scumi replied:
Agreed, updated the code using the new NMEAFormatter.

MaxKellermann commented on src/Formatter/NMEAFormatter.cpp:
```diff
@@ -0,0 +1,117 @@
+               "GPRMC,%02u%02u%02u,%c,%02u%02u.%03u,%c,%03u%02u.%03u,%c,%05.1f,%05.1f,%02u%02u%02u,%05.1f,%c",
+               now.hour, now.minute, now.second,
+               info.location_available ? 'A' : 'V',
+               (int)location.latitude.AbsoluteDegrees(),
```
Does this suffer from rounding errors? Your way of implementing this looks suspicious - it uses different methods to extract degrees and the fractional part; to guarantee that it's free from rounding errors, only one method should be used. The IGC2NMEA code uses `Angle::ToDMS()` to extract all components.

Scumi replied:
Will use `Angle::ToDMM()` as it has all necessary components ready to use.

MaxKellermann commented:
I just saw that the code I complained about was not actually written by you, but you copied it from the CaiLNav driver. Copying code isn't good! Copying bad code is even worse!
Taking advantage of existing code is good, but instead of pretending you wrote it, the first commit should have just moved the code out of the CaiLNav driver as-is, as bad as it is, with all the mistakes (it is clear then that it's not your fault the code is bad, and on the other hand you don't claim being the author); and then follow-up commits can add more features or fix mistakes.

Scumi commented:
Okay, I restructure the commits to reflect that.

MaxKellermann commented on src/Formatter/NMEAFormatter.cpp:
```diff
@@ -23,14 +23,16 @@ FormatGPRMC(char *buffer, size_t buffer_size, const NMEAInfo &info) noexcept
                (double)info.track.Degrees(),
-               now.day, now.month, now.year % 100);
+               now.day, now.month, now.year % 100,
+               info.variation.AbsoluteDegrees(),
```
Must not use this field without checking its validity

MaxKellermann commented:
I don't understand the commit "Formatter/NMEAFormatter: add $ and Checksum directly in the formatter" - this adds complexity, but the commit message doesn't justify that additional complexity. And it changes the timeout to a different value.

Scumi commented:
> I don't understand the commit "Formatter/NMEAFormatter: add $ and Checksum directly in the formatter" - this adds complexity, but the commit message doesn't justify that additional complexity. And it changes the timeout to a different value.

Yes, the timeout should be the same as for `PortWriteNMEA()`. I will fix that.

I think the formatter should spit out a "ready-to-use" NMEA sentence and that also includes the starting $ and ends with a checksum. If you disagree I can remove that part, but will mention in the code comments that this might need to be added by anyone calling the formatter.

MaxKellermann commented:
I'm not sure what's the best API; it was left out because we have the PortWriteNMEA() function for that, so it was natural to omit the checksum in NMEA line generators, and add it only while sending; only the serial port transfer needs to be protected. There was no explanation at all in your commit message, and I still don't know why we should change that and not use PortWriteNMEA(). Certainly, PortWriteNMEA() isn't the best code design, but if we change it, we must be confident that the new code is better, and I'm not confident here.

Scumi commented:
Ok, fair enough. Will remove this part then and update the other commits. The IGC2NMEA-code looked so nice and clean with this.

MaxKellermann commented:
I like how simple the final patch to add the actual feature has become!
